modules/database/src/std/filters/ts.c

```diff
@@ -56,34 +162,248 @@ static db_field_log* filter(void* pvt, dbChannel *chan, db_field_log *pfl) {
+    }
+    pfl->no_elements = 1;
+    pfl->type = dbfl_type_val;
+
+    if (func(pvt, pfl)) {
+        db_delete_field_log(pfl);
+        pfl = NULL;
+    }
+
+    return pfl;
+}
+
+static void ts_to_array(tsPrivate const *settings,
+                        epicsTimeStamp const *ts,
+                        epicsUInt32 arr[2]) {
+    arr[0] = ts->secPastEpoch;
+    arr[1] = ts->nsec;
+    if (settings->epoch == tsEpochUnix) {
+        /* Cannot use epicsTimeToWhatever because Whatever uses signed ints */
+        arr[0] += POSIX_TIME_AT_EPICS_EPOCH;
+    }
+}
+
+static int ts_seconds(tsPrivate const *settings, db_field_log *pfl) {
+    epicsUInt32 arr[2];
+    ts_to_array(settings, &pfl->time, arr);
+    pfl->field_type = DBF_ULONG;
+    pfl->field_size = sizeof(epicsUInt32);
+    pfl->u.v.field.dbf_ulong = arr[0];
+    return 0;
+}
+
+static int ts_nanos(tsPrivate const *settings, db_field_log *pfl) {
+    epicsUInt32 arr[2];
+    ts_to_array(settings, &pfl->time, arr);
+    pfl->field_type = DBF_ULONG;
+    pfl->field_size = sizeof(epicsUInt32);
+    pfl->u.v.field.dbf_ulong = arr[1];
+    return 0;
+}
+
+static int ts_double(tsPrivate const *settings, db_field_log *pfl) {
+    epicsUInt32 arr[2];
+    ts_to_array(settings, &pfl->time, arr);
+    pfl->field_type = DBF_DOUBLE;
+    pfl->field_size = sizeof(epicsFloat64);
+    pfl->u.v.field.dbf_double = arr[0] + arr[1] * 1e-9;
+    return 0;
+}
+
+static int ts_array(tsPrivate const *settings, db_field_log *pfl) {
+    pfl->field_type = DBF_ULONG;
+    pfl->field_size = sizeof(epicsUInt32);
+    pfl->type = dbfl_type_ref;
+    pfl->u.r.pvt = NULL;
+    pfl->u.r.field = allocTsArray();
+    if (pfl->u.r.field) {
+        pfl->no_elements = 2;
+        pfl->dtor = freeTsArray;
+        ts_to_array(settings, &pfl->time, (epicsUInt32*)pfl->u.r.field);
+    } else {
+        pfl->no_elements = 0;
+        pfl->dtor = NULL;
+    }
+    return 0;
+}
+
+static int ts_string(tsPrivate const *settings, db_field_log *pfl) {
+    char const *fmt;
+    char *field;
+    size_t n;
+
+    switch (settings->str) {
+    case tsStringEpics:
+        fmt = "%Y-%m-%d %H:%M:%S.%06f";
+        break;
+    case tsStringIso:
+        fmt = "%Y-%m-%dT%H:%M:%S.%06f%z";
+        break;
+    case tsStringInvalid:
+    default:
+        logicErrorMessage();
+        return 1;
+    }
+
+    pfl->field_type = DBF_STRING;
+    pfl->field_size = MAX_STRING_SIZE;
+    pfl->type = dbfl_type_ref;
+    pfl->u.r.pvt = NULL;
+    pfl->u.r.field = allocString();
+
+    if (!pfl->u.r.field) {
+        pfl->no_elements = 0;
+        pfl->dtor = NULL;
+        return 0;
+    }
+
+    pfl->dtor = freeString;
+
+    field = (char *)pfl->u.r.field;
+    n = epicsTimeToStrftime(field, MAX_STRING_SIZE, fmt, &pfl->time);
+    if (!n) {
+        field[0] = 0;
+    }
+
+    return 0;
+}
+
+static db_field_log *filter(void *pvt, dbChannel *chan, db_field_log *pfl) {
+    (void)chan;
+
+    tsPrivate *settings = (tsPrivate *)pvt;
```

Comment:
This fails with older compilers which insist that C code have local definitions before statements.

Looking more closely, the appveyor builds didn't run for some reason.  @exzombie could you rebase to pick up any appveyor config changes?

From a vs2012 build:

```
ts.c(284) : error C2275: 'tsPrivate' : illegal use of this type as an expression
        ts.c(68) : see declaration of 'tsPrivate'
ts.c(284) : error C2065: 'settings' : undeclared identifier
ts.c(286) : error C2065: 'settings' : undeclared identifier
ts.c(286) : error C2223: left of '->mode' must point to struct/union
ts.c(318) : error C2275: 'tsPrivate' : illegal use of this type as an expression
        ts.c(68) : see declaration of 'tsPrivate'
ts.c(318) : error C2065: 'settings' : undeclared identifier
ts.c(319) : error C2065: 'settings' : undeclared identifier
ts.c(319) : error C2223: left of '->mode' must point to struct/union
ts.c(329) : error C2275: 'tsPrivate' : illegal use of this type as an expression
        ts.c(68) : see declaration of 'tsPrivate'
ts.c(329) : error C2065: 'settings' : undeclared identifier
ts.c(331) : error C2065: 'settings' : undeclared identifier
ts.c(331) : error C2223: left of '->mode' must point to struct/union
ts.c(348) : error C2065: 'settings' : undeclared identifier
ts.c(348) : error C2223: left of '->mode' must point to struct/union
ts.c(379) : error C2275: 'tsPrivate' : illegal use of this type as an expression
        ts.c(68) : see declaration of 'tsPrivate'
ts.c(379) : error C2065: 'settings' : undeclared identifier
ts.c(381) : error C2065: 'settings' : undeclared identifier
ts.c(381) : error C2223: left of '->mode' must point to struct/union
ts.c(381) : error C2223: left of '->epoch' must point to struct/union
ts.c(381) : error C2223: left of '->str' must point to struct/union
ts.c(385) : warning C4113: 'void *(__cdecl *)()' differs in parameter lists from 'void *(__cdecl *)(void)'
```

Reply:
Ugh, I'll try to remember that, thanks. I prefer to treat release notes separately because they *always* have conflicts. Anyway, when I noticed that Appveyor jobs aren't running, I had them run overnight [against my fork](https://ci.appveyor.com/project/exzombie/epics-base) instead, and they were all successful.

Reply:
Thanks. We just discussed that Appveyor problem in today's Core group meeting and will be removing the documentation directory from Appveyor's `skip_commits.files` setting to avoid this in the future.